## NEAT: NGS pipelines made easy

NEAT is a next generation analysis toolkit that supports the analysis of large data including metagene analysis (VhIPseq) and differential gene expression analysis (RNAseq). 
NEAT runs on NGS data produced by ChIPpip and RNApip packages downloadable on GitHub.
NEAT can be run either on a cluster via the command line or directly via the available applescript wrapper. This allows users to generate metagene analysis (for ChIPseq data) as well as differentially regulated gene calling (for RNAseq data) using a simple double click approach. All files including count tables, RPKM values, DEG, venn diagrams, feature-centered enrichment plots, smear plots, etc are automatically saved and archived for a userfriendly and reproducable flow. 
The NEAT toolkit can easily be implemented in any institution with limited to no programming knowledge.
NEAT has been developped in collaboration with wet-lab scientists as well as bioinformaticiens to insure userfriendliness, management of complicated experimental setups and reproducibility in the big data era.
To start using NEAT, please read the README file.


### What does NEAT do
NEAT has two main tools for each RNAseq and ChIPseq project:

[ 1 ]       Download  -  Download RNA/ChIPpip project from remote sever to local computer

[ 2 ]       Analysis  -  Metegene analysis, deifferentiatial gene expression ,etc


### Tutorial
Tutorials can be found in the Vignette folder. This accompanies new users through a the entire analysis of a test dataset provided as part of the package.

